blog/posts/dune-universal-libraries-preview.md

```diff
@@ -0,0 +1,241 @@
+codebase that can render React.js components on the server, such that the
+[Ahrefs](https://ahrefs.com) website can be rendered on the server without
+JavaScript. And, finally, having React.js hydrate the server-rendered HTML in
+the browser. [Dave](https://twitter.com/davesnx) explains the motivation behind
+this goal in more depth [in his blog
+](https://sancho.dev/blog/server-side-rendering-react-in-ocaml).
+
+To look at a specific example, we'll start with a Melange codebase already
+using [`reason-react`](https://github.com/reasonml/reason-react). Our goal is
+to get those `reason-react` components to compile server-side with the OCaml
+compiler, where we'll use
+[`server-reason-react`](https://github.com/ml-in-barcelona/server-reason-react)
+as a drop-in replacement for `reason-react`.
+
+
+What gets in our way is that:
+
+- not everything is supported on both sides: some Melange modules use APIs that
+  don't exist in OCaml (and extensive shimming is undesirable).
+  - vice-versa on the Melange side; especially code that calls into C bindings.
+- we can't vary what implementation to use inside a module or conditionally
+  apply different preprocessing steps and/or flags.
+
+In summary, we would like to to vary specific module implementations across the
+same library, based on their compilation target. If we try to use it in a
+real-world codebase, we'll also find the need to vary preprocessing
+definitions, compilation flags, the set of modules belonging to the library –
+effectively most fields in the `(library ..)` stanza.
+
+## A First ~~Hack~~ Approach
+
+We concluded that it would be desirable to write two library definitions. That
+would allow us to configure each `(library ..)` stanza field separately,
+achieving our goal.
+
+But Dune doesn't allow you to have two libraries with the same name. How could
+it? If Dune derives the artifact directory for libraries from their `(name ..)`
+field, two conflicting names compete for the same artifact directory.
+
+So we first tried to work around that, and set up:
+
+- unwrapped (`(wrapped false)`) Dune libraries with different names
+    - with unwrapped libraries, we could share modules across compilation
+      targets, e.g. `react.ml` originating from both `reason-react` and
+      `server-reason-react`;
+- defined in different directories;
+- `(copy_files ..)` from one of the directories into the other, duplicating
+  shared modules.
+    - Modules with the same name and different implementations, specifc to each
+      directory.
+
+```clj
+;; native/dune
+(library
+ (name native_lib)
+ (wrapped false)
+ (modules a b c))
+
+;; melange/dune
+(library
+ (name melange_lib)
+ (wrapped false)
+ (modes melange)
+ (modules a b c))
+
+;; Copy modules `A` and `B` from `../native`
+(copy_files# ../native
+ (files {a,b}.ml{,i}))
+
+;; module `C` has a specific Melange implementation
+(rule
+ (with-stdout-to c.ml
+  (echo "let backend = \"melange\"")))
```

Comment:
> I think the key part missing in this example dune config block is the native/shared folder where we copy_files from. Then one can use include_subdirs unqualified, and put platform specific modules in either native or melange.

Isn't this satisfied by the `copy_files#` stanzas below?

Reply:
because creating a rule makes it explicit that it's a melange-specific and native-specific module. The alternative would be to describe an external dune file, but this way we can just show dune files and nothing else.

Reply:
Makes sense.